File: world.ex
defmodule World do
  @callback hash(wrld :: term()) :: term()
  @callback start() :: term()
  @callback is_target(wrld :: term()) :: term()
  @callback is_dead(wrld :: term()) :: term()
  @callback br(wrld :: term()) :: term()

  require Logger

  def is_overload() do
    p_limit = :erlang.system_info(:process_limit)
    p_count = :erlang.system_info(:process_count)
    g_limit = p_limit * 0.9
    p_count > g_limit
  end

  def std_hash(world),
    do: :erlang.phash2(world)

  # хранилище миров - стандартное
  # def lib(mod), do: lib(mod, [])
  # def lib(mod, list) do
  #   receive do
  #     :stop -> :ok
  #     {wrld, pid} ->
  #       w_hash = apply(mod, :hash, [wrld])
  #       new_list = case :lists.member(w_hash, list) do
  #         false ->
  #           send(pid, :ok)
  #           [w_hash] ++ list
  #         true ->
  #           send(pid, :exist)
  #           list
  #       end
  #       lib(mod, new_list)
  #     _ -> :ok
  #   end
  # end

  # хранилище миров - ets
  def lib(mod) do
    tbl = :ets.new(:wrld_list, [])
    lib(mod, tbl)
  end
  def lib(mod, tbl) do
    receive do
      :stop -> :ok
      {wrld, pid} ->
        w_hash = apply(mod, :hash, [wrld])
        case :ets.member(tbl, w_hash) do
          false ->
            send(pid, :ok)
            :ets.insert(tbl, {w_hash})
          true ->
            send(pid, :exist)
        end
        lib(mod, tbl)
      _ -> :ok
    end
  end

  # на вход подается список кортежей миров, список делится на первый мир в списке Wrld
  # и на хвост Tail из оставшихся
  def ent(_mod, []), do: :ok

  def ent(mod, [wrld|tail]) do
    case is_overload() do
      false ->
        try do
            spawn(__MODULE__, :ent, [mod, wrld])
            ent(mod, tail)
        catch
          _ ->
            Logger.error("world:ent spawn overload")
            ent(mod, wrld)
            ent(mod, tail)
        end
      _ ->
        ent(mod, wrld)
        ent(mod, tail)
    end
  end

  # на вход подается кортеж одного мира
  def ent(mod, wrld) do
    # проверяем существует ли такой мир уже
    :global.send('lib_srv', {wrld, self()})
    receive do
      :ok ->
        # является ли целевым
        apply(mod, :is_target, [wrld])
        # является ли тупиком
        is_dead = apply(mod, :is_dead, [wrld])
        # если не тупик - плодим ветки и идем в них
        if not(is_dead) do
          new_branches = apply(mod, :br, [wrld])
          ent(mod, new_branches)
        else
          :ok
        end
      :exist ->
        :ok
    end
  end

  def start(mod, start_world) do
    Logger.info("World.start mod=#{inspect(mod)}")
    Logger.info("World.start start_world=#{inspect(start_world)}")
    pid_lib_srv = spawn(__MODULE__, :lib, [mod])
    :global.re_register_name('lib_srv', pid_lib_srv)
    Logger.info("World.start lib(worldlist) PID=#{inspect(pid_lib_srv)}")
    pid_lib_srv = spawn(__MODULE__, :ent, [mod, start_world])
    Logger.info("World.start ent PID=#{inspect(pid_lib_srv)}")
  end
end

File: world.erl
-module(world).
%world framework functions
-callback hash(Wrld :: term()) -> term().
-callback start() -> term().
-callback is_target(Wrld :: term()) -> term().
-callback is_dead(Wrld :: term()) -> term().
-callback br(Wrld :: term()) -> term().

-export([start/2, std_hash/1]).

-export([lib/1, lib/2, ent/2]).
%-export([srch/3]).
-export([is_overload/0]).

std_hash(Wrld)-> erlang:phash2(Wrld).

is_overload() ->
	PLimit = erlang:system_info(process_limit),
	PCount = erlang:system_info(process_count),
	GLimit = PLimit * 0.9,
	PCount > GLimit.

%хранилище миров - стандартное
%lib(Mod) -> lib(Mod, []).
% 
%lib(Mod, List) ->
%    receive
%		stop -> ok;
%		{Wrld, Pid} ->
%			WHash = Mod:hash(Wrld),
%			NewList = case lists:member(WHash, List) of
%						false ->
%							Pid ! ok,
%							[WHash]++List;
%						true ->
%							Pid ! exist,
%							List
%					  end,
%			lib(Mod, NewList);
%		_ -> ok
 %   end.

%%хранилище миров - ets с процессами (переделать под spawn overload)
%srch(Mod, Tbl, {Wrld, Pid}) ->
%	WHash = Mod:hash(Wrld),
%	case ets:member(Tbl, WHash) of
%		false ->
%			io:format("ets:insert(Tbl=~w, ~w)~n", [Tbl, WHash]),
%			ets:insert(Tbl, {WHash, ok}),
%			Pid ! ok;							
%		true ->
%			Pid ! exist
%   end.
%
%lib(Mod) -> 
%	Tbl = ets:new(wrld_list, []),
%	lib(Mod, Tbl).
% 
%lib(Mod, Tbl) ->
%    receive
%		stop -> ok;
%		{Wrld, Pid} ->
%			%without exception
%			%spawn(?MODULE, srch_proc, [Mod, Tbl, {Wrld, Pid}, self()]),
%			
%			case is_overload() of
%				false ->
%					try spawn(?MODULE, srch, [Mod, Tbl, {Wrld, Pid}]) of
%						_Pid -> ok
%					catch
%						_:_ -> 
%							io:format("world:lib spawn overload~n", []),
%							srch(Mod, Tbl, {Wrld, Pid})
%					end;
%				_ ->
%					%io:format("world:lib soft overload~n", []),
%					srch(Mod, Tbl, {Wrld, Pid})
%			end,
%			lib(Mod, Tbl);
%		_ -> ok
%	end.
	
%%хранилище миров - ets
lib(Mod) -> 
	Tbl = ets:new(wrld_list, []),
	lib(Mod, Tbl).
 
lib(Mod, Tbl) ->
    receive
		stop -> ok;
		{Wrld, Pid} ->
			WHash = Mod:hash(Wrld),
			case ets:member(Tbl, WHash) of
				false ->
					Pid ! ok,
					ets:insert(Tbl, {WHash});
				true ->
					Pid ! exist
			end,
			lib(Mod, Tbl);
		_ -> ok
    end.
	
%%на вход подается список кортежей миров, список делится на первый мир в списке Wrld
%%и на хвост Tail из оставшихся
ent(_Mod, []) ->
	%io:format("ent(_Mod, []) -> ok~n", []),
	ok;

ent(Mod, [Wrld|Tail]) ->
	%io:format("ent(~w, [~w|~w]) ->~n", [Mod, Wrld, Tail]),
	
	case is_overload() of
		false ->
			try spawn(?MODULE, ent, [Mod, Wrld]) of
			%%erlang pool
			%%try pool:pspawn(?MODULE, ent, [Mod, Wrld]) of
				_Pid -> ent(Mod, Tail)
			catch
				_:_ -> 
				io:format("world:ent spawn overload~n", []),
				ent(Mod, Wrld),
				ent(Mod, Tail)
			end;
		_ ->
			ent(Mod, Wrld),
			ent(Mod, Tail)
	end;
	

%%на вход подается кортеж одного мира
ent(Mod, Wrld) ->
	%io:format("ent(~w, ~w) ->~n", [Mod, Wrld]),
    %% проверяем существует ли такой мир уже
	global:send("lib_srv", {Wrld, self()}),
    receive 
        ok ->
			%io:format("world:ent World not exist~n", []),
            Mod:is_target(Wrld), %% является ли целевым
            Is_dead = Mod:is_dead(Wrld), %% является ли тупиком
            if
                (Is_dead==false) -> %% если не тупик - плодим ветки и идем в них
                    NewBranches = Mod:br(Wrld),
                    ent(Mod, NewBranches);
                true -> ok
            end;
        exist ->
			%io:format("world:ent World exist~n", []),
            ok
    end.

%%запуск расчета
start(Mod, StartWrld) ->
	io:format("world:start(~w,~w)~n", [Mod, StartWrld]),
    PID_lib_srv = spawn(?MODULE, lib, [Mod]),
	global:register_name("lib_srv", PID_lib_srv),
	io:format("world:start lib PID=~w~n", [PID_lib_srv]),
	PID_ent = spawn(?MODULE, ent, [Mod, StartWrld]),
	io:format("world:start ent PID=~w~n", [PID_ent]).
